library/rbtrees.pl

```diff
@@ -141,29 +141,31 @@
 rb_empty(t(Nil,Nil)) :-
     Nil = black('',_,_,'').
 
-%!  rb_lookup(+Key, -Value, +Tree) is semidet.
+%!  rb_lookup(+Key, ?Value, +Tree) is semidet.
 %
 %   True when Value is associated with Key   in the Red-Black tree Tree.
 %   The given Key may include variables, in   which  case the RB tree is
-%   searched for a key with equivalent,   as  in (==)/2, variables. Time
+%   searched for a key with equivalent  variables (using (==)/2).   Time
 %   complexity is O(log N) in the number of elements in the tree.
+%   Throws an error if Key is not sufficiently instantiated (see rb_in/3
+%   if you wish to backtrack over all keys).
 
 rb_lookup(Key, Val, t(_,Tree)) =>
     lookup(Key, Val, Tree).
 
-lookup(_, _, black('',_,_,'')) :- !, fail.
-lookup(Key, Val, Tree) :-
+lookup(Key, _Val, black('',_,_,'')), nonvar(Key) => fail.
```

Comment:
rbtrees (and assoc) do allow for unbound keys.   This breaks code.

Reply:
Python requires that keys be effectively "const" - that is, they must have a hash function that doesn't change (this is enforced by a convention that mutable types shouldn't have hash functions ... Prolog, of course, is not mutable (except for `setarg/3` et al) but having a variable in the term makes it act like a mutable value when used as a key.

A variable in a key won't behave the way most people expect (even if they read the comment about `(==)/2` being used and not `(=)/2`). I've removed the "nonvar" checks for `rb_lookup/3`, `rb_update/*`, `rb_delete/*` `ord_list_to_rbtree/2`, but I'm not happy with that (I've left them as "%TODO: nonvar(...)"). My view is that variables in the key that participate in a comparison are an error waiting to happen.

I've changed the comment in `rb_lookup/3` and others to strongly advise against variables in keys. I've also marked the tests that I think should throw an error with "TODO"s.

Reply:
nonvar/1 is not good enough.   You either need ground/1 or verify that ?=/2 holds between each pair of keys.   Oops, ?=/2 isn't good enough either because a(X,1) and a(Y,2) satisfy ?=/2, but may be reordered using unification.   Similar issues applies to sort/2, @</2 and anything that relies on standard order of terms.   The ordsets library is pretty useful for effectively computing e.g. the set if shared variables between two (large) terms but relies on the same assumptions.

I see little reason to advise against relying on standard order for non-ground terms.   Instead I prefer to point users at the fact that unification may change ordering and therefore one should either understand the consequences or ensure all keys are ground.   Possibly this should be a section in the manual one can refer to.

Reply:
The `nonvar` tests have already been removed, replaced by "TODO"s. I'll add a comment to the code about those TODOs -- as you say, it would be even better to document the subtle problems.

For what it's worth, Python doesn't prevent creating a hash function for a mutable structure and makes no attempt to detect that a key's hash has changed in a Python "dict". This seems to be acceptable behavior in the Python community, so the similar Prolog problem of keys containing variables should also be something that we don't need to check. (Python will throw an error if a (mutable) structure without a hash function is used as a key with a Python "dict" -- in Prolog, `term_hash(foo(X),Hash)` succeeds with `Hash` remaining uninstantiated ... is this intended behavior?)

BTW, I did the original changes from ":-" to "=>" because of a typo in my code (which the singleton variables check didn't catch) that resulted in using an empty tree; I then though it might be useful to at least do a partial check for dangerous keys but doing this properly seems to be too complicated.

Reply:
I've changed the TODOs in commit f52057b.

An earlier commit had already added this warning to rb_lookup, rb_update, rb_delete:

> This predicate may fail or give unexpected results if Key is not sufficiently instantiated - see `rb_in/3` for backtracking  over keys.